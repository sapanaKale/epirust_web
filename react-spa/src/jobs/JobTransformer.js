/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

const extractStatus = (simulations) => {
  if(simulations.some(s => s.status === 'failed')) return 'failed'
  if(simulations.some(s => s.status === 'running')) return 'running'
  if(simulations.every(s => s.status === 'finished')) return 'finished'
  if(simulations.some(s => s.status === 'finished')) return 'running'
  return 'in-queue'
}

export const reduceStatus = (data) => {
  return {
    ...data,
    status: extractStatus(data.simulations)
  }
}
