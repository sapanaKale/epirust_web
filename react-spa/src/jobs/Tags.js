/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import PropTypes from "prop-types";

export default function Tags({tags = [], isSmall = false}) {
  const classes = ['tags', ...(isSmall ? ['small'] : [])].join(' ');
  return (
    <ul data-testid="tags" className={classes}>
      {tags.map(tag =>
        <li key={tag.id} data-testid={`tag-${tag.id}`}>{tag.name}</li>
      )}
    </ul>
  );
}

const TagType = PropTypes.shape({id: PropTypes.string, name: PropTypes.string});
Tags.propTypes = {
  tags: PropTypes.arrayOf(TagType).isRequired,
  isSmall: PropTypes.bool
};
