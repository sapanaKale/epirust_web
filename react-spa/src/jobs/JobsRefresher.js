/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {get} from "../common/apiCall";
import {reduceStatus} from "./JobTransformer";
import {LOADING_STATES} from "../common/constants";

export class JobsRefresher {
  constructor(updateJobs, updateLoadingState) {
    this.updateJobs = updateJobs;
    this.updateLoadingState = updateLoadingState;
  }

  #refreshJobs = (jobsToProcess) => {
    const jobsToFetch = jobsToProcess.filter(j => j.status !== "finished")
    if (jobsToFetch.length > 0) {
      const url = `/jobs?jobIds=${jobsToFetch.map(j => j._id).join(",")}`
      get(url)
        .then(res => res.json())
        .then(jobsToUpdate => {
          this.updateJobs(prevJobs => {
            const updatedJobs = prevJobs.map(pj => {
              const jobToUpdate = jobsToUpdate.find(ju => ju._id === pj._id);
              return jobToUpdate ? reduceStatus(jobToUpdate) : pj;
            });

            setTimeout(() => this.#refreshJobs(updatedJobs), 15000);
            return updatedJobs;
          })
        })
    }
  }

  start = () => {
    get("/jobs")
      .then(res => res.json())
      .then(receivedJobs => {
        const convertedJobs = receivedJobs.map(reduceStatus).reverse();
        this.updateJobs(convertedJobs)
        this.updateLoadingState(LOADING_STATES.FINISHED)
        setTimeout(() => this.#refreshJobs(convertedJobs), 15000)
      })
      .catch(() => this.updateLoadingState(LOADING_STATES.FAILED))
  }
}