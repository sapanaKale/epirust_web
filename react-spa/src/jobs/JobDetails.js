/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import "../styles/components/jobs/_job-details.scss"
import {NavItem} from "../common/NavItem";
import {Redirect, Route, Switch} from "react-router-dom";
import GridPage from "../grid";
import PropTypes from 'prop-types';
import {TimeSeries} from "../time-series";
import Tags from "./Tags";

export const JobDetails = ({jobId, config}) => {
  const linkPrefix = `/jobs/${jobId}`;
  const gridEnabled = config.enable_citizen_state_messages;

  function renderGraphTabs() {
    return (
      <div className="row">
        <div className="col-8">
          <ul className="nav nav-tabs nav-fill">

            <NavItem name="Time Series" linksTo={`${linkPrefix}/time-series`} activeOnExactMatch={true}/>
            {gridEnabled && <NavItem name="Grid" linksTo={`${linkPrefix}/grid`}/>}
            <NavItem name="Config" linksTo={`${linkPrefix}/config`}/>

          </ul>
        </div>

        <div className="col-4">
          <div className="job-meta">
            <p>{`Job: `}<code>{jobId}</code></p>
            <Tags tags={config.tags}/>
          </div>
        </div>
      </div>
    );
  }

  function renderContentForTab() {
    return (
      <div className='job-details-content'>
        <Switch>

          <Route exact path={"/jobs/:id/time-series"}>
            <TimeSeries jobId={jobId}/>
          </Route>

          <Route exact path={"/jobs/:id/grid"}>
            {gridEnabled ? <GridPage jobId={jobId}/> : <Redirect to={`/jobs/${jobId}/time-series`}/>}
          </Route>

          <Route exact path={"/jobs/:id/config"}>
            {<pre>{JSON.stringify(config, undefined, 4)}</pre>}
          </Route>
        </Switch>
      </div>
    );
  }


  return (
    <div className="job-details">
      {renderGraphTabs()}
      {renderContentForTab()}
    </div>
  )
};

JobDetails.propTypes = {
  jobId: PropTypes.string.isRequired,
  config: PropTypes.object.isRequired
};
