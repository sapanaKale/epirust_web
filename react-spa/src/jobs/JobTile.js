/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import '../styles/components/jobs/_job.scss'
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import Tags from "./Tags";

export const JobTile = ({jobId, status, tags = [], isActive = false, onClick}) => {
  const statusText = {
    "running": "In-Progress",
    "finished": "Finished",
    "failed": "Failed",
    "in-queue": "In-Queue"
  };

  return (
    <li className="list-group-item borderless item-less-padding">
      <div onClick={() => onClick(jobId)} className={`simulation-tab ${isActive ? "active shadow" : ""} ${status}`}>
        <div className="card-body card-body-less-padding">
          <p className="card-title title"><code>{jobId}</code></p>
          <p data-testid={`job-status-${jobId}`} className="job-status">{statusText[status]}</p>
          <Tags isSmall={true} tags={tags}/>
        </div>
      </div>
    </li>
  );
};

JobTile.propTypes = {
  jobId: PropTypes.string.isRequired,
  status: PropTypes.oneOf(["running", "finished", "failed", "in-queue"]).isRequired,
  isActive: PropTypes.bool,
  tags: PropTypes.arrayOf(PropTypes.object)
};
