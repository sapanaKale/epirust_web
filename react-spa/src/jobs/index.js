/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import '../styles/components/jobs/_jobs-list.scss'
import {LOADING_STATES} from "../common/constants";
import {JobsContainer} from "./JobsContainer";
import {LoadingComponent} from "../common/LoadingComponent";
import {JobsRefresher} from "./JobsRefresher";

export const JobsView = () => {
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING);
  const [jobs, updateJobs] = useState([]);

  useEffect(() => {
    const jobsRefresher = new JobsRefresher(updateJobs, updateLoadingState);
    jobsRefresher.start()
  }, []);

  return <LoadingComponent loadingState={loadingState}>
    <JobsContainer jobs={jobs}/>
  </LoadingComponent>;
};
