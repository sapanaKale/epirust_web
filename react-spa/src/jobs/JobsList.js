/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {JobTile} from "./JobTile";
import React, {useEffect, useState} from "react";
import TagsSelect from "../simulation-configuration/TagsSelect";
import {useHistory, useLocation} from "react-router-dom";

const REACT_SELECT_ACTIONS = {
  CLEAR: 'clear',
  SELECT: 'select-option',
  POP: 'pop-value',
  REMOVE: 'remove-value'
};


export const JobsList = ({jobs, activeJob}) => {
  let history = useHistory();
  const location = useLocation();

  const [tagIdsToFilter, updateTagIdsToFilter] = useState([]);
  const [listHeight, updateListHeight] = useState(0);
  const filterFormRef = React.useRef(null);

  useEffect(() => {
    if (!location)
      return;

    let filterTags = location.search
      .match(/filterTags=.+?(?=&|$)/);

    if (!filterTags)
      return;

    const filterTagsUrl = filterTags && filterTags[0]
      .split("=")[1]
      .split(',');

    updateTagIdsToFilter(filterTagsUrl)

  }, [location]);

  function handleSelectChange(changeValues, {action}) {
    const changedIds = changeValues ? changeValues.map(option => option.value) : [];

    const filterQueryParams = changedIds.length > 0
      ? `?filterTags=${changedIds.join(',')}`
      : "";

    switch (action) {
      case REACT_SELECT_ACTIONS.SELECT:
      case REACT_SELECT_ACTIONS.CLEAR:
      case REACT_SELECT_ACTIONS.POP:
      case REACT_SELECT_ACTIONS.REMOVE:
        updateTagIdsToFilter(changedIds);
        history.push(`${location.pathname}${filterQueryParams}`);
        break;
    }
  }

  useEffect(() => {
    if (!filterFormRef)
      return;

    const parentElementHeight = filterFormRef.current.parentElement.clientHeight,
      formHeight = filterFormRef.current.clientHeight;

    updateListHeight(parentElementHeight - formHeight)
  }, [filterFormRef, tagIdsToFilter]);

  function isJobTagged(job) {
    if (!tagIdsToFilter || !tagIdsToFilter.length)
      return true;
    return job.config.tags.some(({id}) => tagIdsToFilter.includes(id));
  }

  function handleJobClick(jobId) {
    const filterQueryParams = tagIdsToFilter.length > 0
      ? `?filterTags=${tagIdsToFilter.join(',')}`
      : "";

    history.push(`/jobs/${jobId}${filterQueryParams}`)
  }

  return (
    <>
      <form ref={filterFormRef} className="job-filter-controls" onSubmit={e => e.preventDefault()}>
        <TagsSelect
          label={"Filter"}
          placeholder={"Type tag name..."}
          onChange={handleSelectChange}
          tagIdsSelected={tagIdsToFilter}/>
      </form>
      <ul className="list-group scrollable job-list" data-testid="job-list" style={{maxHeight: listHeight}}>
        {jobs
          .filter(isJobTagged)
          .map(job =>
            <JobTile
              key={job._id}
              jobId={job._id}
              status={job.status}
              tags={job.config.tags}
              isActive={activeJob ? job._id === activeJob._id : false}
              onClick={handleJobClick}
            />)
        }
      </ul>
    </>
  );
};
