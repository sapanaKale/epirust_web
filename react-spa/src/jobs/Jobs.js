/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {JobsList} from "./JobsList";
import {JobDetails} from "./JobDetails";
import React from "react";

export const Jobs = ({jobs, activeJob}) => {
  return (
    <div className="row jobs-page">
      <div className="col-2 jobs">
        <JobsList jobs={jobs} activeJob={activeJob}/>
      </div>
      <div className="col-10 left-border scrollable details">
        {activeJob ?
          <JobDetails jobId={activeJob._id} config={activeJob.config}/>
          : <div>Invalid job id</div>}
      </div>
    </div>
  )
};
