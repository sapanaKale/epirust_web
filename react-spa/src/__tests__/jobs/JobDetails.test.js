/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {JobDetails} from "../../jobs/JobDetails";
import {render} from "@testing-library/react";
import React from "react";
import {MemoryRouter} from "react-router-dom";
import {act} from "react-dom/test-utils";


let defaultConfig = {enable_citizen_state_messages: false, tags: []};

function getComponent(jobId, config = defaultConfig) {
  return <JobDetails jobId={jobId} config={config}/>
}

describe('Job Details', () => {
  it('should render the time series if the view is time-series', async () => {
    const component = render(<MemoryRouter initialEntries={['/jobs/123/time-series']}>
      {getComponent('123')}
    </MemoryRouter>);


    //TODO: Find a better way to assert. Changes irrelevant to the test would fail this test
    expect(component.container).toMatchSnapshot()
  });

  it('should render the config', async () => {
    const component = render(<MemoryRouter initialEntries={['/jobs/123/config']}>
      {getComponent('123')}
    </MemoryRouter>);

    expect(component.container).toMatchSnapshot()
  });

  it('should render loader initially when no info received from socket', async () => {

    const {asFragment} = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
      {getComponent('123')}
    </MemoryRouter>);


    expect(asFragment()).toMatchSnapshot()
  });

  it('should render grid if grid is enabled in config', async () => {
    const component = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
      {getComponent('123', {...defaultConfig, enable_citizen_state_messages: true})}
    </MemoryRouter>);

    expect(component.container).toMatchSnapshot()
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });
});
