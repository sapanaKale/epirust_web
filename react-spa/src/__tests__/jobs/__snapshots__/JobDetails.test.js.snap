// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Details should render grid if grid is enabled in config 1`] = `
<div>
  <div
    class="job-details"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/grid"
            >
              Grid
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <div
          class="job-meta"
        >
          <p>
            Job: 
            <code>
              123
            </code>
          </p>
          <ul
            class="tags"
            data-testid="tags"
          />
        </div>
      </div>
    </div>
    <div
      class="job-details-content"
    >
      <div
        class="grid-wrap"
      >
        <div
          class="multi-spinners"
          data-testid="loader"
          id="loader"
        >
          <div
            class="spinner-grow text-dark"
          />
          <div
            class="spinner-grow text-dark"
          />
          <div
            class="spinner-grow text-dark"
          />
          <div
            class="spinner-grow text-dark"
          />
          <div
            class="spinner-grow text-dark"
          />
          <div
            class="spinner-grow text-dark"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Job Details should render loader initially when no info received from socket 1`] = `
<DocumentFragment>
  <div
    class="job-details"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <div
          class="job-meta"
        >
          <p>
            Job: 
            <code>
              123
            </code>
          </p>
          <ul
            class="tags"
            data-testid="tags"
          />
        </div>
      </div>
    </div>
    <div
      class="job-details-content"
    >
      <div
        class="multi-spinners"
        data-testid="loader"
        id="loader"
      >
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Job Details should render the config 1`] = `
<div>
  <div
    class="job-details"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <div
          class="job-meta"
        >
          <p>
            Job: 
            <code>
              123
            </code>
          </p>
          <ul
            class="tags"
            data-testid="tags"
          />
        </div>
      </div>
    </div>
    <div
      class="job-details-content"
    >
      <pre>
        {
    "enable_citizen_state_messages": false,
    "tags": []
}
      </pre>
    </div>
  </div>
</div>
`;

exports[`Job Details should render the time series if the view is time-series 1`] = `
<div>
  <div
    class="job-details"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <div
          class="job-meta"
        >
          <p>
            Job: 
            <code>
              123
            </code>
          </p>
          <ul
            class="tags"
            data-testid="tags"
          />
        </div>
      </div>
    </div>
    <div
      class="job-details-content"
    >
      <div
        class="multi-spinners"
        data-testid="loader"
        id="loader"
      >
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
        <div
          class="spinner-grow text-dark"
        />
      </div>
    </div>
  </div>
</div>
`;
